fix: Use strategies for workspace plugins.

kinyoklion opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/release-please/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1978,  #2089 and maybe #2109 🦕

The main goal here is to generate a base pull request using the strategy defined for the release. Then add to that the updates that are needed specifically to the workspace. This will result in generating the correct changelog header as well as updating "extra-files".

I need to implement tests for this yet, but I wanted to get some opinions on the approach before piling too much code on. I've only made changes for the node workspace, but the concept could be extended to to the other workspace plugins.

Example pr created with these changes showing fixes for #1978 #2089:
https://github.com/kinyoklion/release-please-monorepo-tests/pull/10/files

kinyoklion commented on src/plugins/node-workspace.ts:
```diff
@@ -271,20 +278,31 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
+    const strategy = this.strategiesByPath[updatedPackage.location];
+    const latestRelease = this.releasesByPath[updatedPackage.location];
+
+    const basePullRequest = strategy
```
Make a base pull request that contains the "extra-files" updaters as well as the changelog header.

kinyoklion commented on src/plugins/node-workspace.ts:
```diff
@@ -297,16 +315,17 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
-              dependencyNotes,
-              this.logger
-            ),
+            versionHeaderRegex: `\n###? v?[${version
```
The base PR will include the version. The default regex would insert changes above the last version. Which makes it backward.

I need to come back and cleanup the "append" updater if the PR moves forward.

kinyoklion commented on src/strategies/base.ts:
```diff
@@ -254,19 +254,19 @@ export abstract class BaseStrategy implements Strategy {
     draft?: boolean,
-    labels: string[] = []
+    labels: string[] = [],
+    bumpOnlyOptions?: BumpReleaseOptions
```
I needed some indicator to bypass all the early out conditions associated with no commits or changelog entries.

An alternate approach to this whole thing may be to generate all the PRs always, and then prune ones we don't want.

kinyoklion commented on src/plugins/node-workspace.ts:
```diff
@@ -297,16 +315,17 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
-      headRefName: BranchName.ofTargetBranch(this.targetBranch).toString(),
+      headRefName:
+        basePullRequest?.headRefName ??
+        BranchName.ofTargetBranch(this.targetBranch).toString(),
```
For #2109

kinyoklion commented:
I am working on the CLA situation.

chingor13 commented on src/strategy.ts:
```diff
@@ -47,7 +51,8 @@ export interface Strategy {
     draft?: boolean,
-    labels?: string[]
+    labels?: string[],
+    bumpOnlyOptions?: BumpReleaseOptions
```
I'm a bit unclear on what these options are for. Perhaps a docstring is good enough to describe what it's use is for.

kinyoklion replied:
Yes, I will add some documentation. 

I wasn't 100% sure on this, but basically these serve as an indicator that you are creating a release with a version number update, even if there were no associated conventional commits. 

So it is:
A: An indicator that we want to make a release even though there are no calculated changes.
B: The version number we want for that release.

chingor13 reviewed (changes requested):
Thank you for this and apologies for the late review.

In general I think this is a fine approach, although I don't fully understand the edge cases you are trying to handle.  Test case would demonstrate the primary scenario it's fixing and prevent regressions in the future. Test cases would also demonstrate the edge cases you're trying to handle.

kinyoklion commented:
> Thank you for this and apologies for the late review.
> 
> In general I think this is a fine approach, although I don't fully understand the edge cases you are trying to handle. Test case would demonstrate the primary scenario it's fixing and prevent regressions in the future. Test cases would also demonstrate the edge cases you're trying to handle.

Thank you @chingor13,

Now that I know the general approach is fine I will add some test cases that make this more clear.

The situation is not really an edge case. If you have any workspace dependent packages, then it doesn't generate change logs correctly or update files with version numbers.

Imagine you have two packages. Package A and package B. Package B depends on package A.

You commit a change which affects files in package A: "feat: Add my feature."

The release PR that gets generated will have an incorrect changelog. It will just append a "dependencies" section to the top of the changelog.

Example:
![image](https://github.com/googleapis/release-please/assets/4955475/1d983dc0-564c-403d-9d73-d849926c6a34)

The second major problem is that if package B had any `extra-files`, then the version numbers in those files would not be updated.

Both of these are because version bumps only from dependencies do not run any of the steps associated with the base strategy being used.

Thank you,
Ryan

kinyoklion commented on __snapshots__/node-workspace.js:
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
+
+<details><summary>@here/pkgB: 2.2.3</summary>
+
+## [2.2.3](https://github.com/googleapis/node-test-repo/compare/pkgB-v2.2.2...pkgB-v2.2.3) (1983-10-10)
```
This line here represents one of the primary differences. Without this change, there would be no header line here. Both in the PR and in the Changelog.

kinyoklion commented on __snapshots__/node-workspace.js:
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
+other notes
+`
+
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 3'] = `
```
This demonstrates the correct header line being added to the changelog.

kinyoklion commented on __snapshots__/node-workspace.js:
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
+`
+
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 5'] = `
+## 1.1.2 (2023-12-15)
```
This also would not have been present. This doesn't include a diff, because it didn't have the previous version information included.

kinyoklion commented on test/plugins/node-workspace.ts:
```diff
@@ -372,6 +376,171 @@ describe('NodeWorkspace plugin', () => {
+      expect(nodeCandidate).to.not.be.undefined;
+      const updates = nodeCandidate!.pullRequest.updates;
+
+      assertHasUpdate(updates, 'node2/my-file', Generic);
```
Without invoking the base strategy we would not have the update for extra-files.

kinyoklion commented:
I've added some additional doc comments, added tests, and made some comments in the PR for those tests.

In this PR I am only focusing on node, but I believe that these same problems likely affect the other workspace plugins.

kinyoklion commented:
I found some situation where the changelog is still not correct, so I am investigating that and will update.

TimothyJones commented:
I am certain this would fix #2109. 

In addition to the issues you mentioned, I think this _might_ also fix #1946, or at least address one of the issues that contributes to it.
